### PR TITLE
Phase 2: Tandem-Focused and Novel Approaches (8 parallel experiments)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The tandem split (36.7 Pa) is 2-3x worse than other splits (13-18 Pa). This gap represents the largest untapped improvement potential. Tandem airfoils involve wake interactions between two bodies — a fundamentally harder physics problem. This PR attacks the tandem gap from multiple angles: architecture specialization, augmentation, physics constraints, and Transolver++ techniques from the literature.

## Instructions

**CRITICAL: All experiments must change these constants at the top of train.py:**
```python
MAX_TIMEOUT = 180.0  # 3 hours
MAX_EPOCHS = 500
n_layers = 2         # in model_config
```

**Common schedule for all (n_layers=2):**
- warmup=20, T_max=180, eta_min=5e-5
- ema_start=120, temp_anneal>=140
- Progressive volume ramp: epoch < 80
- Tandem curriculum: epoch < 20
- Noise annealing: epoch / 120

### GPU 0: Extended tandem curriculum (30 epochs)
```python
# Extend tandem curriculum from 10 → 30 epochs
# Change: if epoch < 10 → if epoch < 30
# Hypothesis: tandem needs more warmup before the model is ready to learn wake interactions
# Also increase tandem boost clamp: max(1.0, min(6.0, ...)) instead of max 4.0
```
`CUDA_VISIBLE_DEVICES=0 python train.py --wandb_name "nezuko/p2-tandem-curr-30" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 1: Tandem-specific surface weight boost
```python
# Double the surface weight for tandem samples specifically:
# In the surf_loss computation, add: tandem_surf_multiplier = 2.0
# surf_per_sample *= torch.where(is_tandem_batch, tandem_surf_multiplier, 1.0)
# This is SEPARATE from the existing adaptive_boost
```
`CUDA_VISIBLE_DEVICES=1 python train.py --wandb_name "nezuko/p2-tandem-surf-2x" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 2: Gumbel-Softmax slice assignment (Transolver++ technique)
```python
# Replace softmax in Physics_Attention_Irregular_Mesh with Gumbel-Softmax:
# During training: add Gumbel noise to slice logits before softmax
# gumbel_noise = -torch.log(-torch.log(torch.rand_like(slice_logits) + 1e-8) + 1e-8)
# slice_weights = softmax((slice_logits + gumbel_noise) / temp)
# During eval: use standard softmax (no noise)
# This encourages sharper, more discriminative slice assignments
```
`CUDA_VISIBLE_DEVICES=2 python train.py --wandb_name "nezuko/p2-gumbel-slice" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 3: Adaptive per-point temperature (Transolver++ technique)
```python
# Replace fixed temperature with per-point adaptive temperature:
# Add: self.temp_proj = nn.Linear(dim_head, 1)  # in Physics_Attention_Irregular_Mesh.__init__
# In forward: temp = self.temperature + self.temp_proj(x_mid).squeeze(-1).unsqueeze(-1) * 0.1
# Each mesh point gets its own temperature adjustment based on its features
# This helps differentiate surface vs volume node routing
```
`CUDA_VISIBLE_DEVICES=3 python train.py --wandb_name "nezuko/p2-adaptive-temp" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 4: Mixup augmentation (alpha=0.2, skip tandem)
```python
# After loading batch, before forward pass:
# if model.training and not is_tandem_batch.all():
#   non_tandem = ~is_tandem_batch
#   if non_tandem.sum() >= 2:
#     lam = torch.distributions.Beta(0.2, 0.2).sample()
#     idx = torch.randperm(non_tandem.sum())
#     x[non_tandem] = lam * x[non_tandem] + (1-lam) * x[non_tandem][idx]
#     y_norm[non_tandem] = lam * y_norm[non_tandem] + (1-lam) * y_norm[non_tandem][idx]
# Only mix non-tandem samples to avoid confusing tandem physics
```
`CUDA_VISIBLE_DEVICES=4 python train.py --wandb_name "nezuko/p2-mixup" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 5: Pressure gradient regularization (physics constraint)
```python
# Add a regularization term that penalizes large spatial gradients in predicted pressure:
# After computing pred, for surface nodes:
#   sort surface nodes by x-coordinate
#   compute finite differences of predicted pressure along surface
#   pressure_grad = (pred_surf[1:, 2] - pred_surf[:-1, 2]) / (x_surf[1:, 0] - x_surf[:-1, 0])
#   grad_penalty = (pressure_grad.abs() - threshold).clamp(min=0).mean()
#   loss += 0.01 * grad_penalty
# threshold = 5.0 (allow normal pressure gradients, penalize spikes)
```
`CUDA_VISIBLE_DEVICES=5 python train.py --wandb_name "nezuko/p2-pgrad-reg" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 6: Separate surface encoder head
```python
# Add a second output head specifically for surface nodes:
# self.surf_head = nn.Sequential(nn.Linear(n_hidden, n_hidden), nn.GELU(), nn.Linear(n_hidden, 3))
# In forward: surf_pred = self.surf_head(fx)
# Use surf_pred for surface loss, main output for volume loss
# Interpolate: final_pred = torch.where(is_surface, 0.5*fx_pred + 0.5*surf_pred, fx_pred)
# This gives surface predictions a dedicated pathway
```
`CUDA_VISIBLE_DEVICES=6 python train.py --wandb_name "nezuko/p2-surf-head" --wandb_group "phase2-tandem" --agent nezuko`

### GPU 7: Tandem-aware feature: encode foil interaction explicitly
```python
# Add explicit interaction features for tandem samples:
# For each node, compute distance to the NEAREST point on EACH foil
# foil1_dist = dsdf[:, :4].norm(dim=-1, keepdim=True)
# foil2_dist = dsdf[:, 4:8].norm(dim=-1, keepdim=True)
# interaction_feat = foil1_dist / (foil2_dist + 1e-3)  # ratio encodes relative proximity
# Concatenate interaction_feat to input x (increase fun_dim by 1)
# For non-tandem samples, set interaction_feat = 1.0 (no interaction)
```
`CUDA_VISIBLE_DEVICES=7 python train.py --wandb_name "nezuko/p2-tandem-interact" --wandb_group "phase2-tandem" --agent nezuko`

## Baseline
| Metric | Value |
|--------|-------|
| val/loss | 0.833 |
| p_in | 17.9 Pa |
| p_oodc | 14.0 Pa |
| p_tan | 36.7 Pa |
| p_re | 27.5 Pa |

---

## Research

Consulted researcher-agent on Gumbel-Softmax, adaptive temperature, and pressure gradient regularization before implementing.

**Gumbel-Softmax (GPU 2):** Add noise to raw logits before temperature scaling: `(logits + gumbel_noise) / tau`. Use clamped uniform sampling `u.clamp(1e-7, 1-1e-7)` rather than `rand + 1e-8` to avoid NaN in fp16. Training-only (eval uses standard softmax). No hard=True since we need continuous weights for the einsum aggregation.

**Adaptive temperature (GPU 3):** Zero-init `temp_proj` weight so training starts identical to baseline and learns per-point adjustments gradually. Scale factor 0.1 keeps initial temperature perturbations small (±0.05 around 0.5). Clamp min=0.1 to prevent temperature collapse.

**Pressure gradient regularization (GPU 5):** Researcher flagged that sorting surface nodes by x-coordinate conflates upper and lower airfoil surfaces, creating spurious gradients at leading/trailing edge where surfaces meet. Also that the penalty in normalized prediction space may have very different scale from the intended threshold (5.0) in physical space. This proved fatal — GPU 5 failed completely with 10× worse losses.

## Results

All runs trained for 180 min (wall-clock timeout), reaching ~247–250 epochs. The baseline ran for only 30 min / 1 layer; the ~12% improvement in val/loss seen here is primarily attributable to the longer training and additional layer, not the specific techniques.

### Per-variant results

| Variant | val/loss | p_in | p_oodc | p_tan | p_re | Peak Mem | W&B Run |
|---------|----------|------|--------|-------|------|----------|---------|
| **Baseline** | 0.833 | 17.9 | 14.0 | 36.7 | 27.5 | — | — |
| curr30 (GPU 0) | 0.7472 | 15.4 | 10.7 | 38.3 | 26.0 | 26.6 GB | gkmq7mse |
| surf2x (GPU 1) | 0.7478 | 14.5 | 10.8 | 37.7 | 26.1 | 26.6 GB | 820qz3db |
| **gumbel (GPU 2)** | **0.7273** | 14.9 | 10.6 | **36.1** | 25.7 | 27.4 GB | cfym82ra |
| **adaptive-temp (GPU 3)** | **0.7236** | 15.2 | 10.7 | **36.2** | 25.8 | 27.0 GB | t1f443zg |
| mixup (GPU 4) | 0.7563 | 15.1 | 12.0 | 36.4 | 26.4 | 26.6 GB | wb52hip2 |
| ~~pgrad (GPU 5)~~ | ~~10.727~~ | ~~450.9~~ | ~~224.3~~ | ~~437.9~~ | ~~201.5~~ | 26.9 GB | s0ou7xee |
| surf-head (GPU 6) | 0.7459 | 15.5 | 11.2 | 37.7 | 25.7 | 27.8 GB | lx215vla |
| tandem-interact (GPU 7) | 0.7514 | 16.4 | 11.0 | 37.4 | 26.4 | 26.4 GB | tfb30ois |

### What happened

**Winners: Gumbel-Softmax and Adaptive Temperature** are the top performers on both overall val/loss and tandem pressure MAE. Gumbel achieves 36.1 Pa tandem pressure (vs 36.7 baseline, −1.6%), adaptive-temp achieves 36.2 Pa. These are closely matched. The benefit makes physical sense: tandem samples involve wake interactions where different mesh nodes (wake, leading edge of rear foil, suction surfaces) should be routed to very different physics slices. Sharper, more discriminative slice assignments (Gumbel) and per-point temperature control (adaptive-temp) both help the model specialize routing for these complex configurations.

**Extended tandem curriculum hurt tandem (curr30):** p_tan=38.3, *worse* than baseline. Suppressing tandem gradients for 30 epochs (vs 20) leaves the model less adapted to tandem physics at the end, not more. The warmup is too long relative to total training at this scale.

**Surf2x and surf-head neutral:** Both improve in-dist and OOD splits slightly but don't help tandem specifically. The surf2x multiplier likely over-focuses on already-well-predicted surface regions rather than the harder wake-interaction zones.

**Mixup marginally helpful for tandem (36.4 Pa) but hurts OOD-cond (12.0 vs 10.7):** Interpolating samples smooths out the feature space for non-tandem but doesn't directly address tandem wake physics.

**Tandem-interact feature shows no improvement:** The dsdf ratio (foil1_dist / foil2_dist) provides geometric proximity information, but the model may already be learning this implicitly from the raw dsdf features. The feature adds noise rather than signal at this scale.

**Pgrad completely failed:** The pressure gradient threshold (5.0) was applied to predictions in the *normalized loss space*, not physical Pa. In normalized space, pressure gradients are orders of magnitude larger, so the penalty dominated training from epoch 1. Loss never fell below 10× baseline. This is consistent with the researcher's warning that this approach requires careful space-aware implementation.

### Suggested follow-ups

1. **Gumbel + adaptive-temp combined** — both improve routing independently; combining could compound. Worth a dedicated run.
2. **Temperature annealing for Gumbel** — starting Gumbel noise high and annealing down might give better exploration early then sharper routing late.
3. **Pgrad in physical space** — if pressure gradient regularization is to be retried, it must operate on denormalized predictions in Pa. Use `_phys_denorm` output and a physically meaningful threshold (e.g., 50 Pa/chord-length).
4. **Tandem-specific architecture** — rather than a general interaction feature, a tandem-conditional branch (active only for `is_tandem`) might be more targeted.
5. **curr20 re-check** — since curriculum beyond 20 hurts, confirming that 20 is better than 10 would validate the common schedule change made here.

### CODE REVIEW

**Comment by tcapelle** [comment]
Hey nezuko — I notice your pod is running but I don't see any W&B runs started yet. The other students are already well into their 3-hour experiments.

**Please start your experiments immediately.** Here's a quick start:

1. Set `MAX_TIMEOUT = 180.0` and `MAX_EPOCHS = 500` in train.py
2. Set `n_layers = 2` in `model_config`
3. Start with the simpler experiments first:
   - GPU 0: Extended tandem curriculum (just change `if epoch < 10` to `if epoch < 30`)
   - GPU 1: Tandem surface weight boost (add 2x multiplier for tandem surf_loss)
   
4. Then work on the more complex ones (Gumbel-Softmax, adaptive temp, etc.)

Each experiment needs `CUDA_VISIBLE_DEVICES=X` and unique `--wandb_name "nezuko/p2-..."`.

We have limited time — please get started!

**Comment by tcapelle** [comment]
## Round 1 Results — Tandem-Focused and Novel Approaches

Great results, nezuko! Two of your experiments are in the overall top 3!

### Results Summary

| GPU | Experiment | val/loss | p_in | p_oodc | p_tan | p_re | vs Baseline |
|-----|-----------|---------|------|--------|-------|------|-------------|
| 3 | **adaptive-temp** | **0.726** | 15.3 | 10.7 | **36.2** | 25.7 | **-13%** |
| 2 | **gumbel-slice** | **0.727** | **14.8** | **10.6** | 36.1 | **25.7** | **-13%** |
| 0 | tandem-curr-30 | 0.747 | 15.2 | 10.7 | 38.3 | 26.0 | -10% |
| 6 | surf-head | 0.747 | 15.7 | 11.2 | 37.7 | 25.8 | -10% |
| 1 | tandem-surf-2x | 0.747 | 14.5 | 10.8 | 37.7 | 26.1 | -10% |
| 7 | tandem-interact | 0.751 | 16.4 | 11.0 | 37.3 | 26.4 | -10% |
| 4 | mixup | 0.759 | 14.8 | 12.1 | 36.9 | 26.4 | -9% |
| 5 | pgrad-reg | TBD | — | — | — | — | — |

### Key Findings
1. **Adaptive per-point temperature (0.726) is #2 overall!** The Transolver++ technique works.
2. **Gumbel-Softmax slices (0.727) is #3 overall!** Stochastic exploration of slice assignments helps.
3. These two novel approaches outperformed most schedule/architecture changes — huge validation
4. **Tandem curriculum extension** didn't specifically help tandem (38.3 vs 36.7 baseline)
5. **Mixup** worked but wasn't top-tier
6. **Separate surface head** showed promise but didn't differentiate

### Verdict: MERGE adaptive-temp AND gumbel-slice

Both are novel architectural improvements that clearly work. They should be combined in Round 2.

**Comment by tcapelle** [comment]
Closing: Round 1 results captured and incorporated into Round 2 PRs (#1706-#1710). The findings from this PR have been used to design the next round of experiments. Thank you for your excellent work!

**Comment by tcapelle** [comment]
## Round 1 Review — Tandem-Focused Approaches

Excellent work, nezuko. The Transolver++ techniques produced the strongest results here.

**Winners:**
- **Adaptive temperature** (val/loss=0.724, p_tan=36.2) — per-point temperature control works!
- **Gumbel-Softmax** (val/loss=0.727, p_tan=36.1) — stochastic slice exploration helps tandem

**Key insight:** Both routing-improvement techniques (Gumbel, adaptive-temp) beat the architectural approaches (surf-head, tandem-interact). This confirms that better slice assignment is the key lever for tandem improvement.

**Dead ends:**
- **Pressure gradient regularization** — catastrophic failure due to normalized-vs-physical space mismatch
- **Extended curriculum (30 epochs)** — too much suppression hurts tandem

These findings are incorporated into Round 2:
- Gumbel-Softmax tested in frieren's combined experiments (#1706)
- Adaptive temperature tested in frieren (#1706) and nezuko's Round 2 (#1709)
- Combined Gumbel+adaptive-temp tested in nezuko Round 2

**Closing** — results captured, techniques promoted to Round 2. The insight about routing > architecture for tandem is guiding Round 2 design.
